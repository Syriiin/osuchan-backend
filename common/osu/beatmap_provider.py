import os
import urllib.request
from abc import ABC, abstractmethod
from typing import Type, Union

from django.conf import settings
from django.utils.module_loading import import_string


class BeatmapNotFoundException(Exception):
    pass


class AbstractBeatmapProvider(ABC):
    @abstractmethod
    def get_beatmap_file(self, beatmap_id: str) -> str:
        raise NotImplementedError()


class LiveBeatmapProvider(AbstractBeatmapProvider):
    def get_beatmap_file(self, beatmap_id: str) -> str:
        beatmap_path = os.path.join(settings.BEATMAP_CACHE_PATH, str(beatmap_id))

        if not os.path.isfile(beatmap_path):
            beatmap_url = f"{settings.BEATMAP_DL_URL}{beatmap_id}"
            urllib.request.urlretrieve(beatmap_url, beatmap_path)
            if os.path.getsize(beatmap_path) == 0:
                os.remove(beatmap_path)
                raise BeatmapNotFoundException(
                    f"Beatmap {beatmap_id} not found at {beatmap_url}"
                )

        return beatmap_path


class StubBeatmapProvider(AbstractBeatmapProvider):
    def get_beatmap_file(self, beatmap_id: str) -> Union[str, None]:
        beatmap_path = os.path.join(
            os.path.dirname(__file__), "stubdata", "beatmap_provider", str(beatmap_id)
        )

        if not os.path.exists(beatmap_path):
            raise BeatmapNotFoundException(
                f"Beatmap {beatmap_id} not found at {beatmap_path}"
            )

        return beatmap_path


BeatmapProvider: Type[AbstractBeatmapProvider] = import_string(
    settings.BEATMAP_PROVIDER_CLASS
)
