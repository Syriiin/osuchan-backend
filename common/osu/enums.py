# osu! related enums

from enum import IntEnum, StrEnum

# Mods
# https://github.com/ppy/osu-api/wiki#mods


class Mods(IntEnum):
    NONE = 0
    NOFAIL = 1
    EASY = 2
    TOUCH_DEVICE = 4
    HIDDEN = 8
    HARDROCK = 16
    SUDDEN_DEATH = 32
    DOUBLETIME = 64
    RELAX = 128
    HALFTIME = 256
    NIGHTCORE = 512
    FLASHLIGHT = 1024
    AUTO = 2048
    SPUN_OUT = 4096
    AUTOPILOT = 8192
    PERFECT = 16384
    KEY_4 = 32768
    KEY_5 = 65536
    KEY_6 = 131072
    KEY_7 = 262144
    KEY_8 = 524288
    FADE_IN = 1048576
    RANDOM = 2097152
    CINEMA = 4194304
    TARGET_PRACTICE = 8388608
    KEY_9 = 16777216
    KEY_COOP = 33554432
    KEY_1 = 67108864
    KEY_2 = 134217728
    KEY_3 = 268435456
    SCORE_V2 = 536870912
    MIRROR = 1073741824

    KEY_MOD = (
        KEY_1 | KEY_2 | KEY_3 | KEY_4 | KEY_5 | KEY_6 | KEY_7 | KEY_8 | KEY_9 | KEY_COOP
    )
    FREEMOD_ALLOWED = (
        NOFAIL
        | EASY
        | HIDDEN
        | HARDROCK
        | SUDDEN_DEATH
        | FLASHLIGHT
        | FADE_IN
        | RELAX
        | AUTOPILOT
        | SPUN_OUT
        | KEY_MOD
    )
    SCORE_INCREASING = HIDDEN | HARDROCK | DOUBLETIME | FLASHLIGHT | AUTOPILOT | FADE_IN

    SPEED_CHANGING = DOUBLETIME | HALFTIME | NIGHTCORE
    MAP_CHANGING = SPEED_CHANGING | HARDROCK | EASY
    UNRANKED = (
        RELAX | AUTO | AUTOPILOT | KEY_1 | KEY_2 | KEY_3 | KEY_COOP | RANDOM | SCORE_V2
    )


class NewMods(StrEnum):
    NOFAIL = "NF"
    EASY = "EZ"
    TOUCH_DEVICE = "TD"
    HIDDEN = "HD"
    HARDROCK = "HR"
    SUDDEN_DEATH = "SD"
    DOUBLETIME = "DT"
    RELAX = "RX"
    HALFTIME = "HT"
    NIGHTCORE = "NC"
    FLASHLIGHT = "FL"
    AUTO = "AUTO"
    SPUN_OUT = "SO"
    AUTOPILOT = "AP"
    PERFECT = "PF"
    KEY_4 = "4K"
    KEY_5 = "5K"
    KEY_6 = "6K"
    KEY_7 = "7K"
    KEY_8 = "8K"
    FADE_IN = "FI"
    RANDOM = "RN"
    CINEMA = "CN"
    TARGET_PRACTICE = "TP"
    KEY_9 = "9K"
    KEY_COOP = "COOP"
    KEY_1 = "1K"
    KEY_2 = "2K"
    KEY_3 = "3K"
    SCORE_V2 = "V2"
    MIRROR = "MI"
    CLASSIC = "CL"
    BLINDS = "BL"
    ACCURACY_CHALLENGE = "AC"
    DAYCORE = "DC"
    TRACING = "TC"
    MUTED = "MU"
    NO_SCOPE = "NS"


# Gamemodes


class Gamemode(IntEnum):
    STANDARD = 0
    TAIKO = 1
    CATCH = 2
    MANIA = 3


# Beatmap Status


class BeatmapStatus(IntEnum):
    GRAVEYARD = -2
    WIP = -1
    PENDING = 0
    RANKED = 1
    APPROVED = 2
    QUALIFIED = 3
    LOVED = 4
