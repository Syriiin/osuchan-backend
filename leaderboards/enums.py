# leaderboards related enums

from enum import IntEnum


class LeaderboardAccessType(IntEnum):
    GLOBAL = 0
    PUBLIC = 1
    PUBLIC_INVITE_ONLY = 2
    PRIVATE = 3
