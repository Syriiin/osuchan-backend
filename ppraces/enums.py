class PPRaceStatus:
    """Enum for PPRace status."""

    LOBBY = "lobby"
    WAITING_TO_START = "waiting_to_start"
    IN_PROGRESS = "in_progress"
    FINISHED = "finished"
