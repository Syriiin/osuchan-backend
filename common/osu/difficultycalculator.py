from abc import ABC, abstractmethod
from contextlib import AbstractContextManager
from importlib import metadata
from typing import Iterable, NamedTuple, Optional, Type

import httpx
import oppaipy
import rosu_pp_py
from django.conf import settings
from django.utils.module_loading import import_string
from oppaipy.oppaipy import OppaiError

from common.osu.beatmap_provider import BeatmapNotFoundException, BeatmapProvider
from common.osu.enums import Gamemode

OPPAIPY_VERSION = metadata.version("oppaipy")
ROSUPP_VERSION = metadata.version("rosu_pp_py")

# TODO: lazy load this instead of doing at import
difficalcy_osu_info = httpx.get(f"{settings.DIFFICALCY_OSU_URL}/api/info").json()

DIFFICALCY_OSU_ENGINE = difficalcy_osu_info["calculatorPackage"]
DIFFICALCY_OSU_VERSION = difficalcy_osu_info["calculatorVersion"]


class Score(NamedTuple):
    beatmap_id: str
    mods: Optional[int] = None
    count_100: Optional[int] = None
    count_50: Optional[int] = None
    count_miss: Optional[int] = None
    combo: Optional[int] = None


class Calculation(NamedTuple):
    difficulty: float
    performance: float


class DifficultyCalculatorException(Exception):
    pass


class CalculatorClosedException(DifficultyCalculatorException):
    pass


class InvalidBeatmapException(DifficultyCalculatorException):
    pass


class CalculationException(DifficultyCalculatorException):
    pass


class NotYetCalculatedException(DifficultyCalculatorException):
    pass


class AbstractDifficultyCalculator(AbstractContextManager, ABC):
    def __init__(self):
        self.closed = False

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self._close()

    @abstractmethod
    def _close(self):
        raise NotImplementedError()

    def close(self):
        self.closed = True
        self._close()

    def calculate_score(self, score: Score) -> Calculation:
        self._reset()
        self.set_beatmap(score.beatmap_id)

        if score.mods is not None:
            self.set_mods(score.mods)
        if score.count_100 is not None or score.count_50 is not None:
            self.set_accuracy(score.count_100 or 0, score.count_50 or 0)
        if score.count_miss is not None:
            self.set_misses(score.count_miss)
        if score.combo is not None:
            self.set_combo(score.combo)

        self.calculate()
        return Calculation(
            difficulty=self.difficulty_total, performance=self.performance_total
        )

    def calculate_score_batch(self, scores: Iterable[Score]) -> list[Calculation]:
        return [self.calculate_score(score) for score in scores]

    @abstractmethod
    def _reset(self):
        raise NotImplementedError()

    @abstractmethod
    def set_beatmap(self, beatmap_id: str) -> None:
        raise NotImplementedError()

    @abstractmethod
    def set_accuracy(self, count_100: int, count_50: int) -> None:
        raise NotImplementedError()

    @abstractmethod
    def set_misses(self, count_miss: int) -> None:
        raise NotImplementedError()

    @abstractmethod
    def set_combo(self, combo: int) -> None:
        raise NotImplementedError()

    @abstractmethod
    def set_mods(self, mods: int) -> None:
        raise NotImplementedError()

    @abstractmethod
    def _calculate(self) -> None:
        raise NotImplementedError()

    def calculate(self) -> None:
        if self.closed:
            raise CalculatorClosedException(
                "calculate() cannot be called on a closed calculator"
            )

        self._calculate()

    @property
    @abstractmethod
    def difficulty_total(self) -> float:
        raise NotImplementedError()

    @property
    @abstractmethod
    def performance_total(self) -> float:
        raise NotImplementedError()

    @staticmethod
    @abstractmethod
    def engine() -> str:
        raise NotImplementedError()

    @staticmethod
    @abstractmethod
    def version() -> str:
        raise NotImplementedError()

    @staticmethod
    @abstractmethod
    def gamemode() -> Gamemode:
        raise NotImplementedError()


class OppaiDifficultyCalculator(AbstractDifficultyCalculator):
    def __init__(self):
        super().__init__()

        self.oppai_calc = None
        self.beatmap_path = None

    def _close(self):
        if self.oppai_calc is not None:
            self.oppai_calc.close()

    def _reset(self):
        if self.oppai_calc is not None:
            self.oppai_calc.reset()

    def set_beatmap(self, beatmap_id: str) -> None:
        beatmap_provider = BeatmapProvider()
        try:
            self.beatmap_path = beatmap_provider.get_beatmap_file(beatmap_id)
        except BeatmapNotFoundException as e:
            raise InvalidBeatmapException(
                f"An error occured in fetching the beatmap {beatmap_id}"
            ) from e
        self.oppai_calc = oppaipy.Calculator(self.beatmap_path)

    def set_accuracy(self, count_100: int, count_50: int):
        self.oppai_calc.set_accuracy(count_100, count_50)

    def set_misses(self, count_miss: int):
        self.oppai_calc.set_misses(count_miss)

    def set_combo(self, combo: int):
        self.oppai_calc.set_combo(combo)

    def set_mods(self, mods: int):
        self.oppai_calc.set_mods(mods)

    def _calculate(self):
        try:
            self.oppai_calc.calculate()
        except OppaiError as e:
            raise CalculationException(
                f'An error occured in calculating the beatmap "{self.beatmap_path}"'
            ) from e

    @property
    def difficulty_total(self):
        return self.oppai_calc.stars

    @property
    def performance_total(self):
        return self.oppai_calc.pp

    @staticmethod
    def engine():
        return "oppaipy"

    @staticmethod
    def version():
        return OPPAIPY_VERSION

    @staticmethod
    def gamemode():
        return Gamemode.STANDARD


class RosuppDifficultyCalculator(AbstractDifficultyCalculator):
    def __init__(self):
        super().__init__()

        self.rosupp_calc = rosu_pp_py.Performance()
        self.results = None

    def _close(self):
        pass

    def _reset(self):
        self.rosupp_calc = rosu_pp_py.Performance()
        self.results = None

    def set_beatmap(self, beatmap_id: str) -> None:
        beatmap_provider = BeatmapProvider()
        try:
            beatmap_path = beatmap_provider.get_beatmap_file(beatmap_id)
        except BeatmapNotFoundException as e:
            raise InvalidBeatmapException(
                f"An error occured in fetching the beatmap {beatmap_id}"
            ) from e

        try:
            self.beatmap = rosu_pp_py.Beatmap(path=beatmap_path)
        except Exception as e:
            raise InvalidBeatmapException(
                f'An error occured in parsing the beatmap "{self.beatmap_path}"'
            ) from e

    def set_accuracy(self, count_100: int, count_50: int):
        self.rosupp_calc.set_n100(count_100)
        self.rosupp_calc.set_n50(count_50)

    def set_misses(self, count_miss: int):
        self.rosupp_calc.set_misses(count_miss)

    def set_combo(self, combo: int):
        self.rosupp_calc.set_combo(combo)

    def set_mods(self, mods: int):
        self.rosupp_calc.set_mods(mods)

    def _calculate(self):
        self.results = self.rosupp_calc.calculate(self.beatmap)

    @property
    def difficulty_total(self) -> float:
        if self.results is None:
            raise NotYetCalculatedException("Results have not been calculated")
        return self.results.difficulty.stars

    @property
    def performance_total(self) -> float:
        if self.results is None:
            raise NotYetCalculatedException("Results have not been calculated")
        return self.results.pp

    @staticmethod
    def engine():
        return "rosu-pp-py"

    @staticmethod
    def version():
        return ROSUPP_VERSION

    @staticmethod
    def gamemode():
        return Gamemode.STANDARD


class DifficalcyOsuDifficultyCalculator(AbstractDifficultyCalculator):
    def __init__(self):
        super().__init__()

        self.client = httpx.Client()

    def _close(self):
        self.client.close()

    def __difficalcy_score_from_score(self, score: Score) -> dict:
        return {
            k: v
            for k, v in {
                "BeatmapId": score.beatmap_id,
                "Mods": score.mods,
                "Combo": score.combo,
                "Misses": score.count_miss,
                "Mehs": score.count_50,
                "Oks": score.count_100,
            }.items()
            if v is not None
        }

    def calculate_score(self, score: Score) -> Calculation:
        try:
            response = self.client.get(
                f"{settings.DIFFICALCY_OSU_URL}/api/calculation",
                params=self.__difficalcy_score_from_score(score),
            )
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPStatusError as e:
            raise CalculationException(
                f"An error occured in calculating the beatmap {score.beatmap_id}: {e.response.text}"
            ) from e

        return Calculation(
            difficulty=data["difficulty"]["total"],
            performance=data["performance"]["total"],
        )

    def calculate_score_batch(self, scores: Iterable[Score]) -> list[Calculation]:
        try:
            response = self.client.post(
                f"{settings.DIFFICALCY_OSU_URL}/api/batch/calculation",
                json=[self.__difficalcy_score_from_score(score) for score in scores],
            )
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPStatusError as e:
            raise CalculationException(
                f"An error occured in calculating the beatmaps: {e.response.text}"
            ) from e

        return [
            Calculation(
                difficulty=calculation_data["difficulty"]["total"],
                performance=calculation_data["performance"]["total"],
            )
            for calculation_data in data
        ]

    def _reset(self):
        pass

    def set_beatmap(self, beatmap_id: str) -> None:
        raise NotImplementedError()

    def set_accuracy(self, count_100: int, count_50: int) -> None:
        raise NotImplementedError()

    def set_misses(self, count_miss: int) -> None:
        raise NotImplementedError()

    def set_combo(self, combo: int) -> None:
        raise NotImplementedError()

    def set_mods(self, mods: int) -> None:
        raise NotImplementedError()

    def _calculate(self) -> None:
        raise NotImplementedError()

    @property
    def difficulty_total(self) -> float:
        raise NotImplementedError()

    @property
    def performance_total(self) -> float:
        raise NotImplementedError()

    @staticmethod
    def engine() -> str:
        return DIFFICALCY_OSU_ENGINE

    @staticmethod
    def version() -> str:
        return DIFFICALCY_OSU_VERSION

    @staticmethod
    def gamemode():
        return Gamemode.STANDARD


difficulty_calculators: dict[str, type[AbstractDifficultyCalculator]] = {
    name: import_string(calculator_class)
    for name, calculator_class in settings.DIFFICULTY_CALCULATOR_CLASSES.items()
}


def get_difficulty_calculator_class(name: str) -> Type[AbstractDifficultyCalculator]:
    return difficulty_calculators[name]


DifficultyCalculator: Type[AbstractDifficultyCalculator] = import_string(
    settings.DIFFICULTY_CALCULATOR_CLASS
)
