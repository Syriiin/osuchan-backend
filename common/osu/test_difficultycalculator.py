from importlib import metadata
from pathlib import Path

import pytest

from common.osu.difficultycalculator import (
    CalculationException,
    CalculatorClosedException,
    InvalidBeatmapException,
    OppaiDifficultyCalculator,
    RosuppDifficultyCalculator,
)
from common.osu.enums import Mods


class TestOppaiDifficultyCalculator:
    def test_enigne(self):
        assert OppaiDifficultyCalculator.engine() == "oppaipy"

    def test_version(self):
        assert OppaiDifficultyCalculator.version() == metadata.version("oppaipy")

    def test_context_manager(self):
        with OppaiDifficultyCalculator(
            f"{Path(__file__).parent}/testdata/307618.osu"
        ) as calc:
            calc.calculate()
            assert calc.difficulty_total == 4.200401306152344

    def test_invalid_beatmap(self):
        with pytest.raises(CalculationException):
            with OppaiDifficultyCalculator(
                f"{Path(__file__).parent}/testdata/empty.osu"
            ) as calc:
                calc.calculate()

    @pytest.fixture
    def calc(self):
        calc = OppaiDifficultyCalculator(f"{Path(__file__).parent}/testdata/307618.osu")
        calc.calculate()
        return calc

    def test_difficulty_total(self, calc: OppaiDifficultyCalculator):
        assert calc.difficulty_total == 4.200401306152344

    def test_performance_total(self, calc: OppaiDifficultyCalculator):
        assert calc.performance_total == 126.96746063232422

    def test_close(self, calc: OppaiDifficultyCalculator):
        calc.close()
        with pytest.raises(CalculatorClosedException):
            calc.calculate()

    def test_set_accuracy(self, calc: OppaiDifficultyCalculator):
        calc.set_accuracy(14, 1)
        calc.calculate()
        assert calc.difficulty_total == 4.200401306152344
        assert calc.performance_total == 118.77462005615234

    def test_set_misses(self, calc: OppaiDifficultyCalculator):
        calc.set_misses(1)
        calc.calculate()
        assert calc.difficulty_total == 4.200401306152344
        assert calc.performance_total == 123.09051513671875

    def test_set_combo(self, calc: OppaiDifficultyCalculator):
        calc.set_combo(2000)
        calc.calculate()
        assert calc.difficulty_total == 4.200401306152344
        assert calc.performance_total == 106.42977142333984

    def test_set_mods(self, calc: OppaiDifficultyCalculator):
        calc.set_mods(Mods.DOUBLETIME + Mods.HIDDEN)
        calc.calculate()
        assert calc.difficulty_total == 5.919765949249268
        assert calc.performance_total == 397.2521057128906


class TestRosuppDifficultyCalculator:
    def test_enigne(self):
        assert RosuppDifficultyCalculator.engine() == "rosu-pp-py"

    def test_version(self):
        assert RosuppDifficultyCalculator.version() == metadata.version("rosu-pp-py")

    def test_context_manager(self):
        with RosuppDifficultyCalculator(
            f"{Path(__file__).parent}/testdata/307618.osu"
        ) as calc:
            assert calc.difficulty_total == 4.457399442092882

    def test_invalid_beatmap(self):
        with pytest.raises(InvalidBeatmapException):
            RosuppDifficultyCalculator(f"{Path(__file__).parent}/testdata/empty.osu")

    @pytest.fixture
    def calc(self):
        calc = RosuppDifficultyCalculator(
            f"{Path(__file__).parent}/testdata/307618.osu"
        )
        calc.calculate()
        return calc

    def test_difficulty_total(self, calc: RosuppDifficultyCalculator):
        assert calc.difficulty_total == 4.457399442092882

    def test_performance_total(self, calc: RosuppDifficultyCalculator):
        assert calc.performance_total == 135.0350130086423

    def test_close(self, calc: RosuppDifficultyCalculator):
        calc.close()
        with pytest.raises(CalculatorClosedException):
            calc.calculate()

    def test_set_accuracy(self, calc: RosuppDifficultyCalculator):
        calc.set_accuracy(14, 1)
        calc.calculate()
        assert calc.difficulty_total == 4.457399442092882
        assert calc.performance_total == 124.11285312370455

    def test_set_misses(self, calc: RosuppDifficultyCalculator):
        calc.set_misses(1)
        calc.calculate()
        assert calc.difficulty_total == 4.457399442092882
        assert calc.performance_total == 130.64187056414588

    def test_set_combo(self, calc: RosuppDifficultyCalculator):
        calc.set_combo(2000)
        calc.calculate()
        assert calc.difficulty_total == 4.457399442092882
        assert calc.performance_total == 112.78045221645286

    def test_set_mods(self, calc: RosuppDifficultyCalculator):
        calc.set_mods(Mods.DOUBLETIME + Mods.HIDDEN)
        calc.calculate()
        assert calc.difficulty_total == 6.264344677869616
        assert calc.performance_total == 425.80658059182343
