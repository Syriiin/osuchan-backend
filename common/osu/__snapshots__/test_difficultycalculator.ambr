# serializer version: 1
# name: TestDifficalcyCatchDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'total': 5.560488609256414,
      }),
      performance_values=dict({
        'total': 226.4106426420045,
      }),
    ),
  ])
# ---
# name: TestDifficalcyCatchDifficultyCalculator.test_get_beatmap_details
  BeatmapDetails(
    artist='Unknown',
    author='Unknown Creator',
    base_velocity=1.6,
    common_bpm=120,
    difficulty_name='Normal',
    difficulty_settings=dict({
      'approach_rate': 8.3,
      'circle_size': 4,
      'drain_rate': 5,
    }),
    hitobject_counts=dict({
      'banana_showers': 3,
      'fruits': 78,
      'juice_streams': 12,
    }),
    length=45250,
    max_combo=127,
    maximum_bpm=120,
    mininum_bpm=120,
    tick_rate=1,
    title='Unknown',
  )
# ---
# name: TestDifficalcyDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'aim': 2.931964039669495,
        'flashlight': 0,
        'speed': 3.0719986654662823,
        'total': 6.294293342707777,
      }),
      performance_values=dict({
        'accuracy': 83.9838091927015,
        'aim': 122.82365726224828,
        'flashlight': 0,
        'speed': 137.33708223365028,
        'total': 358.86598978801396,
      }),
    ),
    Calculation(
      difficulty_values=dict({
        'aim': 3.1710525152307567,
        'flashlight': 0,
        'speed': 3.0845312684219364,
        'total': 6.55537654432227,
      }),
      performance_values=dict({
        'accuracy': 169.06122268521966,
        'aim': 189.00295549800694,
        'flashlight': 0,
        'speed': 176.71960468663355,
        'total': 556.6053238383237,
      }),
    ),
    Calculation(
      difficulty_values=dict({
        'aim': 3.1710525152307567,
        'flashlight': 0,
        'speed': 3.0845312684219364,
        'total': 6.55537654432227,
      }),
      performance_values=dict({
        'accuracy': 216.31133737582476,
        'aim': 214.36465227080254,
        'flashlight': 0,
        'speed': 209.530750791515,
        'total': 666.2654686196685,
      }),
    ),
    Calculation(
      difficulty_values=dict({
        'aim': 3.2556879022731793,
        'flashlight': 0,
        'speed': 2.0082542663345504,
        'total': 5.854140057199108,
      }),
      performance_values=dict({
        'accuracy': 12.368997155942548,
        'aim': 49.7362338663849,
        'flashlight': 0,
        'speed': 7.181221326403531,
        'total': 74.39657751330903,
      }),
    ),
  ])
# ---
# name: TestDifficalcyDifficultyCalculator.test_context_manager
  list([
    Calculation(
      difficulty_values=dict({
        'aim': 2.118735459664974,
        'flashlight': 0,
        'speed': 2.1684114250452144,
        'total': 4.492310309756154,
      }),
      performance_values=dict({
        'accuracy': 36.07670429437059,
        'aim': 46.26971335416548,
        'flashlight': 0,
        'speed': 49.870242023878504,
        'total': 137.7209357109564,
      }),
    ),
  ])
# ---
# name: TestDifficalcyManiaDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'total': 2.797245912537965,
      }),
      performance_values=dict({
        'difficulty': 43.17076331130473,
        'total': 21.585381655652366,
      }),
    ),
  ])
# ---
# name: TestDifficalcyManiaDifficultyCalculator.test_get_beatmap_details
  BeatmapDetails(
    artist='Unknown',
    author='Unknown Creator',
    base_velocity=1.6,
    common_bpm=120,
    difficulty_name='Normal',
    difficulty_settings=dict({
      'accuracy': 7,
      'drain_rate': 5,
      'key_count': 4,
    }),
    hitobject_counts=dict({
      'hold_notes': 14,
      'notes': 123,
    }),
    length=30500,
    max_combo=151,
    maximum_bpm=120,
    mininum_bpm=120,
    tick_rate=1,
    title='Unknown',
  )
# ---
# name: TestDifficalcyOsuDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'aim': 2.931964039669495,
        'flashlight': 0,
        'speed': 3.0719986654662823,
        'total': 6.294293342707777,
      }),
      performance_values=dict({
        'accuracy': 83.9838091927015,
        'aim': 122.82365726224828,
        'flashlight': 0,
        'speed': 137.33708223365028,
        'total': 358.86598978801396,
      }),
    ),
  ])
# ---
# name: TestDifficalcyOsuDifficultyCalculator.test_get_beatmap_details
  BeatmapDetails(
    artist='senya',
    author='Satellite',
    base_velocity=1.7,
    common_bpm=138,
    difficulty_name='Marathon',
    difficulty_settings=dict({
      'accuracy': 6,
      'approach_rate': 8,
      'circle_size': 4,
      'drain_rate': 5,
    }),
    hitobject_counts=dict({
      'circles': 1093,
      'slider_ticks': 346,
      'sliders': 659,
      'spinners': 1,
    }),
    length=678047,
    max_combo=2758,
    maximum_bpm=168,
    mininum_bpm=121,
    tick_rate=2,
    title='Songs Compilation',
  )
# ---
# name: TestDifficalcyPerformancePlusDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'accuracy': 1.0546478295311437,
        'aim': 2.8804685309467284,
        'flowAim': 1.6727417520197583,
        'jumpAim': 2.8763118223866133,
        'precision': 1.714528074416766,
        'speed': 3.179656468968935,
        'stamina': 2.8045880770770752,
        'total': 6.123075221732318,
      }),
      performance_values=dict({
        'accuracy': 87.89378488351228,
        'aim': 68.38232334186907,
        'flowAim': 13.391867347855435,
        'jumpAim': 68.08670955755021,
        'precision': 14.420763347653594,
        'speed': 99.53445865488443,
        'stamina': 68.30305130767819,
        'total': 259.5756907401086,
      }),
    ),
  ])
# ---
# name: TestDifficalcyPerformancePlusDifficultyCalculator.test_get_beatmap_details
  BeatmapDetails(
    artist='senya',
    author='Satellite',
    base_velocity=1.7,
    common_bpm=138,
    difficulty_name='Marathon',
    difficulty_settings=dict({
      'accuracy': 6,
      'approach_rate': 8,
      'circle_size': 4,
      'drain_rate': 5,
    }),
    hitobject_counts=dict({
      'circles': 1093,
      'slider_ticks': 346,
      'sliders': 659,
      'spinners': 1,
    }),
    length=678047,
    max_combo=2758,
    maximum_bpm=168,
    mininum_bpm=121,
    tick_rate=2,
    title='Songs Compilation',
  )
# ---
# name: TestDifficalcyTaikoDifficultyCalculator.test_calculate_scores
  list([
    Calculation(
      difficulty_values=dict({
        'colour': 0.6552532719640201,
        'rhythm': 0.041101226071239186,
        'stamina': 2.861677137388462,
        'total': 4.616848746682018,
      }),
      performance_values=dict({
        'accuracy': 169.03802944742503,
        'difficulty': 89.1797160815413,
        'total': 275.2953534626361,
      }),
    ),
  ])
# ---
# name: TestDifficalcyTaikoDifficultyCalculator.test_get_beatmap_details
  BeatmapDetails(
    artist='Unknown',
    author='Unknown Creator',
    base_velocity=1.6,
    common_bpm=120,
    difficulty_name='Normal',
    difficulty_settings=dict({
      'accuracy': 7,
      'drain_rate': 5,
    }),
    hitobject_counts=dict({
      'drum_rolls': 30,
      'hits': 200,
      'swells': 8,
    }),
    length=53000,
    max_combo=200,
    maximum_bpm=120,
    mininum_bpm=120,
    tick_rate=1,
    title='Unknown',
  )
# ---
